File: scripts/diagnostics/andromeda_saved_bridge.py
#!/usr/bin/env python3
"""Promote only the reviewed Maldives delta using accepted ANEX identity and exact atoll evidence."""
from __future__ import annotations
import argparse
import hashlib
import json
from pathlib import Path
import re
import unicodedata
from zipfile import ZipFile

OPERATION_ID = 'andromeda-1759-maldives-bridge61-20260910-v1'
PENDING_ZIP_SHA = 'ac0a474ede6a56bc93c7f96c0283b26206e6e6f6f2cb652f90351d47fda25457'
ANEX_ZIP_SHA = '770d183c60ccff1537f1ce33879825f8fae6d40c68e882010d9bde8c49001021'
REQUEST_SHA = '941ad3dccbdd4d4c374bcf572016cc1234a535bfe56a5bcacbf1df99e8eac2b4'

def canonical(value):
    return json.dumps(value,ensure_ascii=False,sort_keys=True,separators=(',',':'),allow_nan=False).encode()

def digest(value):
    return hashlib.sha256(canonical(value)).hexdigest()

def norm(value):
    value=unicodedata.normalize('NFKD',str(value or '')).casefold().replace('ё','е').replace('ı','i')
    value=''.join(c for c in value if not unicodedata.combining(c))
    return ' '.join(re.findall(r'[^\W_]+',value,re.UNICODE))

def name(value):
    value=re.sub(r'\s*\(\s*(?:ex|ех)\s*\.?\s+[^()]+\)\s*$','',str(value or ''),flags=re.I)
    return ' '.join(sorted(w for w in norm(value).split() if w!='hotel'))

def load_zip(path,expected):
    if hashlib.sha256(path.read_bytes()).hexdigest()!=expected:
        raise ValueError('unreviewed_input_archive')
    return ZipFile(path)

def request(pending_zip:Path,anex_zip:Path):
    with load_zip(pending_zip,PENDING_ZIP_SHA) as z:
        saved=json.loads(z.read('maldives.json'))
    with load_zip(anex_zip,ANEX_ZIP_SHA) as z:
        catalog=json.loads(z.read('anex-hotel-catalog-match.json'))['matches']
        accepted={r['anex_hotel_id']:r['catalog_hotel_id'] for r in json.loads(z.read('anex-search-mappings.json'))['rows']}
    if saved['country_id']!=8 or saved['supplier_country_id']!=73 or len(saved['rows'])!=327:
        raise ValueError('saved_country_scope_changed')
    towns={str(t['id']):t for t in saved['towns']}
    index={}
    for r in catalog:
        if r['external_id'] not in accepted or norm(r['country'])!=norm('Мальдивы'):
            continue
        for key in {name(r['name']),name(r['alternate_name'])}-{''}:
            index.setdefault(key,{})[r['external_id']]=r
    rows=[]
    for item in saved['rows']:
        source=item['evidence']['source'];targets=item['targets']
        if item['evidence']['reason']!='geography_unknown' or len(targets)!=1:
            continue
        target=targets[0]
        if norm(target['region_name'])!=norm('Мальдивы') or norm(target['subregion_name']):
            continue
        candidates={}
        for key in {name(source.get('name')),name(source.get('lName'))}-{''}:
            candidates.update(index.get(key,{}))
        if {accepted[a] for a in candidates}!={int(target['id'])}:
            continue
        town=towns.get(str(source.get('townKey')))
        if not town or str(town.get('state'))!='73' or norm(town['name'])!=norm(source.get('town')):
            continue
        bridges=[r for r in candidates.values() if norm(r['town'])==norm(source.get('town'))]
        if not bridges:
            continue
        if hashlib.sha256(item['identity']['evidence_json'].encode()).hexdigest()!=item['evidence_sha256']:
            raise ValueError('original_evidence_mismatch')
        rows.append({'external_hotel_id':str(source['id']),'local_hotel_id':int(target['id']),
            'expected_evidence_sha256':item['evidence_sha256'],'target_name':target['name'],
            'source':source,'official_town':town,
            'anex_bridges':[{'id':r['external_id'],'name':r['name'],'alternate_name':r['alternate_name'],
                'town':r['town'],'country':r['country'],'catalog_hotel_id':accepted[r['external_id']],
                'source_record_sha256':digest(r)} for r in sorted(bridges,key=lambda r:r['external_id'])]})
    rows.sort(key=lambda r:int(r['external_hotel_id']))
    if len(rows)!=61 or len({r['external_hotel_id'] for r in rows})!=61 or len({r['local_hotel_id'] for r in rows})!=61:
        raise ValueError('reviewed_delta_changed')
    payload={'operation_id':OPERATION_ID,'country_id':8,'supplier_country_id':73,
        'saved_capture_sha256':saved['capture_sha256'],'saved_plan_sha256':saved['plan_sha256'],
        'catalog_sha256':saved['catalog_sha256'],'source_pending_sha256':digest(saved),
        'anex_archive_sha256':ANEX_ZIP_SHA,'supplier_calls':0,'rows':rows}
    if digest(payload)!=REQUEST_SHA:
        raise ValueError('reviewed_request_changed')
    return payload

def validate_result(value,payload):
    if (value.get('status')!='accepted' or value.get('operation_id')!=OPERATION_ID
        or value.get('request_sha256')!=digest(payload) or value.get('updated')!=61
        or value.get('readback_verified') is not True or value.get('other_identities_unchanged') is not True
        or value.get('supplier_calls')!=0 or len(value.get('rows',[]))!=61):
        raise ValueError('live_result_unconfirmed')
    expected={r['external_hotel_id']:r['local_hotel_id'] for r in payload['rows']}
    if {str(r['external_hotel_id']):int(r['local_hotel_id']) for r in value['rows']}!=expected:
        raise ValueError('live_pair_mismatch')
    if any(r.get('decision_status')!='accepted' or not re.fullmatch('[0-9a-f]{64}',r.get('evidence_sha256','')) for r in value['rows']):
        raise ValueError('live_evidence_unconfirmed')

def apply(payload,receipt:Path,transport=None):
    from anex_tourvisor_link_import import save
    if receipt.is_symlink():
        raise ValueError('receipt_symlink')
    h=digest(payload)
    if receipt.exists():
        prior=json.loads(receipt.read_bytes())
        if prior.get('state')!='finalized' or prior.get('request_sha256')!=h or prior.get('result_sha256')!=digest(prior.get('result')):
            raise ValueError('reserved_unknown_do_not_replay')
        validate_result(prior['result'],payload)
        return {'status':'already_finalized','new_writes':0,'saved_result':prior['result']}
    save(receipt,{'state':'reserved','operation_id':OPERATION_ID,'request_sha256':h},exclusive=True)
    if transport is None:
        from anex_search3_owner_decisions import ssh_php
        root=Path(__file__).resolve().parents[2]
        source=(root/'app/integrations/anex-search-mapping-registry.php').read_text().removeprefix('<?php')
        source+="\ndefine('CE_LIBRARY_ONLY',true);\n"+Path(__file__).with_name('andromeda_country_expansion.php').read_text().removeprefix('<?php')
        source+='\n'+Path(__file__).with_suffix('.php').read_text().removeprefix('<?php')
        result=ssh_php(source,payload,maximum_bytes=65536)
    else:
        result=transport(payload)
    save(receipt.with_name(receipt.name+'.outcome.json'),result,exclusive=True)
    validate_result(result,payload)
    save(receipt,{'state':'finalized','operation_id':OPERATION_ID,'request_sha256':h,'result_sha256':digest(result),'result':result})
    return result

def main():
    p=argparse.ArgumentParser(description=__doc__)
    p.add_argument('--pending',required=True,type=Path);p.add_argument('--anex',required=True,type=Path)
    p.add_argument('--output',required=True,type=Path);p.add_argument('--apply',action='store_true');p.add_argument('--receipt',type=Path)
    args=p.parse_args()
    if args.apply and args.receipt is None:p.error('--apply requires a durable --receipt')
    payload=request(args.pending,args.anex)
    result=apply(payload,args.receipt) if args.apply else {'status':'prepared_not_applied','request':payload}
    args.output.write_text(json.dumps(result,ensure_ascii=False,sort_keys=True,indent=2)+'\n')
    print(json.dumps(result if args.apply else {'status':result['status'],'rows':len(payload['rows']),'request_sha256':digest(payload)},ensure_ascii=False))

if __name__=='__main__':main()

File: scripts/diagnostics/andromeda_saved_bridge.php
<?php
/** Loaded after the existing registry and CE_LIBRARY_ONLY country helpers. */
const BR_OPERATION='andromeda-1759-maldives-bridge61-20260910-v1';
const BR_REQUEST='941ad3dccbdd4d4c374bcf572016cc1234a535bfe56a5bcacbf1df99e8eac2b4';
function br_payload(array $request): void {
    if(ce_hash($request)!==BR_REQUEST||($request['operation_id']??'')!==BR_OPERATION||count($request['rows']??[])!==61)throw new RuntimeException('reviewed_request_changed');
}
function br_import(PDO $db,array $request,array $data,array $savedPlan): array {
    br_payload($request);
    if((int)$data['country_id']!==8||(int)$data['supplier_country_id']!==73)throw new RuntimeException('country_scope');
    $engine=$db->query("SELECT ENGINE FROM information_schema.TABLES WHERE TABLE_SCHEMA=DATABASE() AND TABLE_NAME='andromeda_hotel_identities'")->fetchColumn();
    if(strtoupper((string)$engine)!=='INNODB')throw new RuntimeException('schema');
    $db->exec('SET SESSION innodb_lock_wait_timeout=10');$db->beginTransaction();$committed=false;
    try {
        $all=$db->query('SELECT * FROM andromeda_hotel_identities ORDER BY supplier_namespace,external_hotel_id LIMIT 100001 FOR UPDATE')->fetchAll(PDO::FETCH_ASSOC);
        if(count($all)>100000)throw new RuntimeException('identity_limit');
        $wanted=[];$anexIds=[];foreach($request['rows'] as $r){$wanted[(string)$r['external_hotel_id']]=$r;foreach($r['anex_bridges'] as $b)$anexIds[(int)$b['id']]=true;}
        $index=[];$preserve=[];foreach($all as $old){$key=$old['supplier_namespace'].':'.$old['external_hotel_id'];$index[$key]=$old;if($old['supplier_namespace']!=='andromeda_catalog'||!isset($wanted[(string)$old['external_hotel_id']]))$preserve[$key]=ce_hash($old);}
        $ids=array_keys($anexIds);sort($ids,SORT_NUMERIC);$marks=implode(',',array_fill(0,count($ids),'?'));
        foreach(['anex_search_hotel_observations','anex_hotel_decisions','anex_hotel_search_mappings','anex_review_pair_exclusions'] as $table){
            $s=$db->prepare('SELECT anex_hotel_id FROM '.$table.' WHERE anex_hotel_id IN ('.$marks.') ORDER BY anex_hotel_id FOR UPDATE');$s->execute($ids);$s->fetchAll();
        }
        $local=ce_local($db,8,true);$current=ce_plan($data,$local);$currentRows=[];$original=[];$hotels=[];$towns=[];
        foreach($current['rows'] as $r)$currentRows[(string)$r['external_hotel_id']]=$r;
        foreach($savedPlan['rows'] as $r)$original[(string)$r['external_hotel_id']]=$r;
        foreach($local['hotels'] as $h)$hotels[(int)$h['id']]=$h;
        foreach($data['catalog']['payload']['TOWNTO'] as $t)$towns[(string)$t['id']]=$t;
        $registry=AnyTourAnexSearchMappingRegistry::fromPdo($db);$updates=[];
        foreach($wanted as $key=>$r){
            $external=(string)$key;$old=$index['andromeda_catalog:'.$external]??null;$saved=$original[$key]??null;
            if(!$old||!$saved||$old['decision_status']!=='pending'||$old['local_hotel_id']!==null||$old['catalog_sha256']!==$request['catalog_sha256']||$old['evidence_sha256']!==$r['expected_evidence_sha256']||hash('sha256',$old['evidence_json'])!==$r['expected_evidence_sha256']||$old['evidence_json']!==$saved['evidence_json'])throw new RuntimeException('stale_identity');
            $prior=json_decode($old['evidence_json'],true,64,JSON_THROW_ON_ERROR);
            $now=json_decode($currentRows[$key]['evidence_json']??'null',true,64,JSON_THROW_ON_ERROR);
            if(($prior['reason']??'')!=='geography_unknown'||($now['candidate_ids']??[])!==[$r['local_hotel_id']]||ce_hash($prior['source'])!==ce_hash($r['source'])||ce_hash($now['source'])!==ce_hash($r['source']))throw new RuntimeException('current_candidate_changed');
            $target=$hotels[$r['local_hotel_id']]??null;
            if(!$target||(int)$target['country_id']!==8||(int)$target['is_active']!==1||$target['name']!==$r['target_name']||ce_norm($target['region_name'])!==ce_norm('Мальдивы')||ce_norm($target['subregion_name'])!=='')throw new RuntimeException('target_geography_changed');
            $source=$r['source'];$town=$towns[(string)$source['townKey']]??null;
            if(!$town||(string)$town['state']!=='73'||(string)$source['stateKey']!=='73'||ce_hash($town)!==ce_hash($r['official_town'])||ce_norm($town['name'])!==ce_norm($source['town'])||ce_norm($source['town'])===ce_norm('Мальдивы'))throw new RuntimeException('official_geography_changed');
            $sourceNames=array_filter([ce_name($source['name']),ce_name($source['lName']??'')]);
            foreach($r['anex_bridges'] as $b){
                if($registry->resolve('anex_online',(string)$b['id'],'preview')!==$r['local_hotel_id']||$b['catalog_hotel_id']!==$r['local_hotel_id']||ce_norm($b['country'])!==ce_norm('Мальдивы')||ce_norm($b['town'])!==ce_norm($source['town'])||!array_intersect($sourceNames,array_filter([ce_name($b['name']),ce_name($b['alternate_name'])])))throw new RuntimeException('anex_bridge_changed');
            }
            $evidence=['prior_evidence'=>$prior,'source'=>'accepted_anex_same_atoll_bridge_20260910','operation_id'=>BR_OPERATION,'request_sha256'=>BR_REQUEST,'anex_bridges'=>$r['anex_bridges'],'official_town'=>$town,'target_name'=>$r['target_name'],'source_pending_sha256'=>$request['source_pending_sha256']];
            $json=ce_json($evidence);$updates[$key]=['external_hotel_id'=>$external,'local_hotel_id'=>$r['local_hotel_id'],'decision_status'=>'accepted','evidence_sha256'=>hash('sha256',$json),'evidence_json'=>$json,'previous_evidence_sha256'=>$r['expected_evidence_sha256']];
        }
        if(count($updates)!==61)throw new RuntimeException('scope');
        $update=$db->prepare("UPDATE andromeda_hotel_identities SET local_hotel_id=?,decision_status='accepted',evidence_sha256=?,evidence_json=? WHERE supplier_namespace='andromeda_catalog' AND external_hotel_id=? AND decision_status='pending' AND local_hotel_id IS NULL AND evidence_sha256=? AND catalog_sha256=?");
        foreach($updates as $r){$update->execute([$r['local_hotel_id'],$r['evidence_sha256'],$r['evidence_json'],$r['external_hotel_id'],$r['previous_evidence_sha256'],$request['catalog_sha256']]);if($update->rowCount()!==1)throw new RuntimeException('conditional_update');}
        $after=ce_existing($db,false);if(count($after)!==count($all))throw new RuntimeException('row_count');foreach($preserve as $key=>$hash)if(($after[$key]??null)!==$hash)throw new RuntimeException('preservation');
        $db->commit();$committed=true;
        $query=$db->prepare("SELECT external_hotel_id,local_hotel_id,decision_status,evidence_sha256 FROM andromeda_hotel_identities WHERE supplier_namespace='andromeda_catalog' AND external_hotel_id=?");$rows=[];
        foreach($updates as $r){$query->execute([$r['external_hotel_id']]);$got=$query->fetch(PDO::FETCH_ASSOC);if(!$got||(string)$got['external_hotel_id']!==$r['external_hotel_id']||(int)$got['local_hotel_id']!==$r['local_hotel_id']||$got['decision_status']!=='accepted'||$got['evidence_sha256']!==$r['evidence_sha256'])throw new RuntimeException('post_commit_readback');$got['external_hotel_id']=(string)$got['external_hotel_id'];$got['local_hotel_id']=(int)$got['local_hotel_id'];$rows[]=$got;}
        $counts=$db->query("SELECT decision_status,COUNT(*) FROM andromeda_hotel_identities WHERE supplier_namespace='andromeda_catalog' GROUP BY decision_status")->fetchAll(PDO::FETCH_KEY_PAIR);
        return ['status'=>'accepted','operation_id'=>BR_OPERATION,'request_sha256'=>BR_REQUEST,'updated'=>61,'readback_verified'=>true,'other_identities_unchanged'=>true,'preserved_count'=>count($preserve),'preservation_sha256'=>ce_hash($preserve),'rows'=>$rows,'current_andromeda_counts'=>$counts,'live_coverage'=>ce_coverage($db),'supplier_calls'=>0];
    }catch(Throwable $e){if(!$committed&&$db->inTransaction())$db->rollBack();throw new RuntimeException($committed?'committed_readback_unknown':'transaction_rolled_back');}
}
if(!defined('BR_LIBRARY_ONLY')){
    error_reporting(0);ob_start();$lock=null;
    try {
        $raw=file_get_contents('php://stdin',false,null,0,150001);if(!is_string($raw)||strlen($raw)>150000)throw new RuntimeException();$request=json_decode($raw,true,64,JSON_THROW_ON_ERROR);br_payload($request);
        $root=realpath(getcwd());if(PHP_SAPI!=='cli'||!$root||basename($root)!=='anytoour.ru')throw new RuntimeException();
        $config=require $root.'/_preview/search3-anex-candidate/.andromeda-private.php';$private=realpath(dirname($config['catalog_path']));if(!$private)throw new RuntimeException();
        $stage=$private.'/'.CE_OPERATION.'-maldives';$data=ce_read($stage.'/capture.json');$plan=ce_read($stage.'/plan.json');$priorResult=ce_read($stage.'/result.json');
        if($priorResult['status']!=='imported'||$priorResult['readback_verified']!==true||ce_hash($data)!==$request['saved_capture_sha256']||ce_hash($plan)!==$request['saved_plan_sha256']||ce_hash($data['catalog'])!==$request['catalog_sha256'])throw new RuntimeException();
        $operation=$private.'/'.BR_OPERATION;$lock=fopen($operation.'.lock','c');if(!$lock||!flock($lock,LOCK_EX))throw new RuntimeException();
        if(is_dir($operation)||!mkdir($operation,0700))throw new RuntimeException('prior_bridge_operation_do_not_replay');
        ce_save($operation.'/reservation.json',['state'=>'reserved','operation_id'=>BR_OPERATION,'request_sha256'=>BR_REQUEST]);
        $helper=realpath($root.(is_file($root.'/data/db-v1.php')?'/data/db-v1.php':'/v2/data/db-v1.php'));if(!$helper||strpos($helper,$root.'/')!==0)throw new RuntimeException();require_once $helper;
        $db=v2_data_db();$db->setAttribute(PDO::ATTR_ERRMODE,PDO::ERRMODE_EXCEPTION);$result=br_import($db,$request,$data,$plan);ce_save($operation.'/result.json',$result);
    }catch(Throwable $e){$result=['status'=>'failed_or_unknown','operation_id'=>BR_OPERATION,'retry'=>false,'supplier_calls'=>0];}
    if($lock){flock($lock,LOCK_UN);fclose($lock);}while(ob_get_level())ob_end_clean();echo ce_json($result);
}
